fastly/data_source_datacenters.go

```diff
@@ -0,0 +1,101 @@
+package fastly
+
+import (
+	"context"
+	"encoding/json"
+	"log"
+	"strconv"
+
+	gofastly "github.com/fastly/go-fastly/v6/fastly"
+	"github.com/fastly/terraform-provider-fastly/fastly/hashcode"
+	"github.com/hashicorp/terraform-plugin-sdk/v2/diag"
+	"github.com/hashicorp/terraform-plugin-sdk/v2/helper/schema"
+)
+
+func dataSourceFastlyDatacenters() *schema.Resource {
+	return &schema.Resource{
+		ReadContext: dataSourceFastlyDatacentersRead,
+
+		Schema: map[string]*schema.Schema{
+			"pops": {
+				Type:        schema.TypeList,
+				Computed:    true,
+				Description: "A list of all Fastly POPs.",
+				Elem: &schema.Resource{
+					Schema: map[string]*schema.Schema{
+						"code": {
+							Type:        schema.TypeString,
+							Computed:    true,
+							Description: "A code representing the POP location.",
+						},
+						"name": {
+							Type:        schema.TypeString,
+							Computed:    true,
+							Description: "The name of the POP.",
+						},
+						"group": {
+							Type:        schema.TypeString,
+							Computed:    true,
+							Description: "A code representing the general region of the world in which the POP location resides.",
+						},
+						"shield": {
+							Type:        schema.TypeString,
+							Computed:    true,
+							Description: "A code representing the shielding name of the POP. The value may be empty if the POP is not available for shielding.",
+						},
+					},
+				},
+			},
+		},
+	}
+}
+
+func dataSourceFastlyDatacentersRead(_ context.Context, d *schema.ResourceData, meta interface{}) diag.Diagnostics {
+
+	conn := meta.(*FastlyClient).conn
+
+	log.Printf("[DEBUG] Reading datacenters")
+
+	datacenters, err := conn.AllDatacenters()
+
+	if err != nil {
+		return diag.Errorf("error fetching datacenters: %s", err)
+	}
+	hashBase, _ := json.Marshal(datacenters)
+	hashString := strconv.Itoa(hashcode.String(string(hashBase)))
+	d.SetId(hashString)
+
+	if err := d.Set("pops", flattenDatacenters(datacenters)); err != nil {
+		return diag.Errorf("error setting datacenters: %s", err)
+	}
+
+	return nil
+
+}
+
+func flattenDatacenters(datacenters []gofastly.Datacenter) []map[string]interface{} {
+
+	pops := make([]map[string]interface{}, len(datacenters))
```

Comment:
If there is no specific need for a slice, then should we maybe create a sized array instead?

```suggestion
	var pops [len(datacenters)]map[string]interface{}
```

Reply:
Hmm, maybe not necessary. 

I was wondering about memory allocations when re-sizing the backing array, but I think that doesn't happen when assigning directly using an index (e.g. only when using `append`, at least from some quick testing it doesn't look like using a slice in this example would cause the backing array to be resized).